### PR TITLE
chore(changelog): new way to maintain the changelog

### PR DESCRIPTION
**Pull request by vm-001**
### Summary

Maintaining changelogs in a single markdown file in the repo is one of the easy ways to maintain and also keep consistency. Since Kong has multiple release versions, sometimes a bugfix needs to be backported to all the supported versions after it gets merged to the master branch. The backport bot is currently broken because of the git conflict. 

I'm Introducing a new way to maintain Kong's changelog, which makes the changelog item become an individual file. The idea is, you don't get the conflict if you don't edit the same file.

Implementation details:
- `CHANGELOG/changelog` is a Lua script that currently provides `preview`, and `release` commands. It only requires Penlight and lyaml libraries.
- Using the template engine(provided by Penglight) which is similar to how Kong renders the nginx configuration.

Notes:
- The changelog entry file is only for demonstration and will be deleted.

### CODE REVIEW

File: CHANGELOG/schema.json
Comment (Guaris) [inline CHANGELOG/schema.json:22]: If you make this a string we may run into issues like:

`plugin` and `plugins` 
Misspelled product names

What if you made this an enum with valid options, and included a `misc` keyword for things that dont fit into the options?

File: CHANGELOG/changelog
Comment (acgoldsmith) [inline CHANGELOG/changelog:15]: Regarding the link to JIRA issue: is this intended for internal audience? We don't reference the JIRA tickets in the docs site changelog.

Comment (vm-001) [reply CHANGELOG/changelog:15]: The CHANGELOG-EE.md include the JIRA ticket in every item, so I was trying to keep the new release note as original as possible. It sounds like you want to use the generated release note to be directly used in the public docs site, I'm happy to try and help. But I think the docs team still needs to review and even edit the generated release note before publishing it to the docs site.


So, in short, questions:
1. Do you think the generated release note should be directly used in the docs site without edit/optimize?
(Currently, the generated released note is more for internal audience)

File: CHANGELOG/schema.json
Comment (vm-001) [reply CHANGELOG/schema.json:22]: Sounds good to me. enum value proposal is welcome. Regarding the `misc`, we can use null to replace it. changelog entry that set the `scope` to null or missing it will be grouped to a default scope.

For example
```
# 1.yaml
message: add a feature
type: feature
pr: 1

# 2.yaml
message: add b feature
type: feature
scope:
pr: 2
```

these two changelog entries will be grouped into a default scope.

**Review by Guaris — APPROVED** [review]

**Comment by hanshuebner** [comment]
Would you consider not requiring sequence numbers in file names?  If order is important, prefixing with a date (`YYYYMMDD`) should do just fine.  Sequence numbers in file names are not easy to maintain and create extra work in cases of conflict.

**Comment by vm-001** [comment]
> Would you consider not requiring sequence numbers in file names? If order is important, prefixing with a date (`YYYYMMDD`) should do just fine. Sequence numbers in file names are not easy to maintain and create extra work in cases of conflict.

The changelog filename can be anything.(e.g. `${pr_number}.yaml` or `a-shot-description-message.yaml`).

In the next phase, I will add a CI workflow to automatically extract the changelog message based on the PR title, and use the PR number as the filename, like `11279.yaml`. Since PR number is already unique, so it does not cause any conflict.

File: CHANGELOG/schema.json
Comment (flrgh) [inline CHANGELOG/schema.json:40]: What do you think about making `pr` an array so that we can link multiple PRs to a changelog entry? It could be handled the same as `issues` or `jiras`.

File: CHANGELOG/changelog-md-template.lua
Comment (flrgh) [inline CHANGELOG/changelog-md-template.lua:43]: It would be good to have ordering of individual entries follow some sort of consistent pattern. The ordering is deliberate in the sections below (breaking changes, deprecations, dependencies, etc...), so individual items _within_ a section should also follow some sort of consistent pattern.

For `scopes` we may want to hard-code the order by some sort of subjective level of significance. Example: core, then plugins, then pdk, and so on. _Or_ we could just `table.sort` the keys alphanumerically.

For entries _within_ a scope, maybe we should sort according to the PR # so that they appear at least _somewhat_ chronologically.

Comment (flrgh) [reply CHANGELOG/changelog-md-template.lua:43]: Other benefit of explicit sorting is that the same input always produces the same output .md file no matter which platform/filesystem/Lua version.

Comment (vm-001) [reply CHANGELOG/changelog-md-template.lua]: LGTM.

The scope is now prioritized.
```
local SCOPE_PRIORITY = { -- smallest on top
  Performance = 10,
  Configuration = 20,
  Core = 30,
  PDK = 40,
  Plugin = 50,
  ["Admin API"] = 60,
  Clustering = 70,
  Default = 100, -- default priority
}
setmetatable(SCOPE_PRIORITY, {
  __index = function()
    return rawget(SCOPE_PRIORITY, "Default") - 1
  end
})
```

Files now added  explicit sorting to avoid the issue that you mentioned.

Regarding the section. I think the section order can be edit as it's now hardcoded.
```
> render_changelog_section("Breaking Changes", enterprise.breaking_changes)
> render_changelog_section("Deprecations", enterprise.deprecations)
> render_changelog_section("Dependencies", enterprise.dependencies)
> render_changelog_section("Features", enterprise.features)
> render_changelog_section("Fixes", enterprise.bugfixes)
```

File: CHANGELOG/schema.json
Comment (vm-001) [reply CHANGELOG/schema.json:40]: Added another field `prs`. But I haven't decided whether to remove the single one `pr` yet.

Comment (flrgh) [reply CHANGELOG/schema.json:40]: I don't have a strong preference, but for simplicity's sake I think I'd vote for removing one of them and just having one field (whether it's `pr` or `prs`).

Relevant idea: you could also make the tool accept either a single item or an array for pr/issue/jira references, so that this:

```yaml
message: my changelog entry
type: feature
scope: core
prs:
  - 123
  - 456
issues: 789
jiras: abc-123
```

...is valid and coerces to this:

```yaml
message: my changelog entry
type: feature
scope: core
prs:
  - 123
  - 456
issues:
  - 789
jiras:
  - abc-123
```

Up to you though--I think things are in good shape here as-is.

Comment (vm-001) [reply CHANGELOG/schema.json]: I removed `pr` to keep things simple first.

**Comment by vm-001** [comment]
Removed test data for get ready for merging

**Comment by fffonion** [comment]
let's also have a PR for EE ready to see how that will look like for EE version backports and ce->ee backports.

**Comment by vm-001** [comment]
> let's also have a PR for EE ready to see how that will look like for EE version backports and ce->ee backports.

Done.

**Comment by vm-001** [comment]
Remove the `skip-changelog` label, and CI fais as expected

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (hanshuebner) [inline .github/PULL_REQUEST_TEMPLATE.md:14]:
```suggestion
- [ ] A changelog file has been added to `CHANGELOG/unreleased/kong`. https://github.com/Kong/kong/CHANGELOG/README.md
```

**Review by hanshuebner — CHANGES_REQUESTED** [review]
It is not clear to me why Lua has been chosen as the implementation language.  It does not seem to be the obvious choice for implementing shell utilities.

In any case, the README.md should be updated so that the examples actually work and there is a description of what a developer needs to do in order to add a new changelog file.  With the examples provided now, all that is created is an unpopulated yaml structure with no information on how to proceed provided.  If the file is meant to be a template, it should contain comments helping the developer to fill it out (Please advise developers to remove the template verbiage and comments).  An interactive utility asking for the fields and guiding through the process would be nice, though.  Also, provide hints on how the files should be named, or even better, suggest a file name automatically.

If this continues to be developed in Lua, please make it work with the virtual environment provided by bazel so that all dependencies are grabbed from there.  The tool seems to work with LuaJIT, and that should be the default and enforced by an appropriate interpreter line in the script.

This error message is insufficient:
```
luajit: CHANGELOG/changelog:186: attempt to index a nil value
stack traceback:
	CHANGELOG/changelog:186: in function 'collect_folder'
	CHANGELOG/changelog:204: in function 'collect_unreleased'
	CHANGELOG/changelog:266: in function 'preview'
	CHANGELOG/changelog:294: in function 'cmd_fn'
	CHANGELOG/changelog:329: in main chunk
	[C]: at 0x0102ea0b4c
```

I will do another review once the appropriate documentation is provided.

**Comment by vm-001** [comment]
> It is not clear to me why Lua has been chosen as the implementation language. It does not seem to be the obvious choice for implementing shell utilities.

Shell/Bash is not the optimal language in terms of YAML parsing and templating. Lua is the main language of this repo, and I suppose all contributors of Kong have Lua knowledge, not to say the core contributors. So using Lua is more easier to maintain compare to the others.


> An interactive utility asking for the fields and guiding through the process would be nice, though. Also, provide hints on how the files should be named, or even better, suggest a file name automatically.

Eventually, the changelog file will be auto added by CI workflow. Ideally, developers shouldn't be distracted by adding the changelog file as it will added by the workflow, so developers/reviewers should focus on editing the generated changelog message to make it proper for end user. In short, I prefer not introce the interactive part in this stage, but may consider at next coming PR.

**Comment by vm-001** [comment]
> This error message is insufficient:
>
> luajit: CHANGELOG/changelog:186: attempt to index a nil value
stack traceback:
	CHANGELOG/changelog:186: in function 'collect_folder'
	CHANGELOG/changelog:204: in function 'collect_unreleased'
	CHANGELOG/changelog:266: in function 'preview'
	CHANGELOG/changelog:294: in function 'cmd_fn'
	CHANGELOG/changelog:329: in main chunk
	[C]: at 0x0102ea0b4c
	
	
What command do you run? It works in my environment.

```
$ luajit CHANGELOG/changelog
Usage: changelog <command> [options]

Commands:
  add <filename> [options]           add a changelog file.
  release <version> [options]        release a release note based on the files in the CHANGELOG/unreleased directory.
  preview <version> [options]        preview a release note based on the files in the CHANGELOG/unreleased directory.

Options:
  -h, --help                         display help for command
  -m, --message (optional string)    changelog message
  -t, --type (optional string)       changelog type
  --folder (string default kong)     which folder under unreleased

Examples:
  changelog preview 1.0.0
  changelog release 1.0.0
  changelog add 1.yaml

```

**Comment by hanshuebner** [comment]
I can't reproduce the error now, but:
```
(kong-dev) mecipt-1 1379_% luajit CHANGELOG/changelog preview 1.0.0 
# 1.0.0

## Kong

(empty lines elided)

(kong-dev) mecipt-1 1379_% 
```
Please update the documentation so that it describes the process of adding a changelog, i.e. the steps that a developer needs to take and how the generation process works.

**Comment by vm-001** [comment]
Could you describe more about
> If this continues to be developed in Lua, please make it work with the virtual environment provided by bazel so that all dependencies are grabbed from there. 

The changelog Lua script only requires `lyaml` and `Penlight` libraries, and they both are defined in `kong-3.5.0-0.rockspec`, which means, not extra lib is required in the Bazel environment. For the non-Bazel environment, I think I can add a Makefile under the /CHANGELOG folder to make it easier to install the lib

```
install_dependencies:
	luarocks install penlight --local
	luarocks install lyaml --local
```

**Comment by hanshuebner** [comment]
> Could you describe more about
> 
> > If this continues to be developed in Lua, please make it work with the virtual environment provided by bazel so that all dependencies are grabbed from there.

I would put `#!/usr/bin/env luajit` into the first line of the `changelog` script and make it executable.  That way, it will work both from the Bazel venv as well as with a locally installed luajit and penlight.

File: CHANGELOG/README.md
Comment (hanshuebner) [inline CHANGELOG/README.md:8]: Let's automate the steps below.  `changelog add` should create the file from the template, possibly asking the user for the PR number.  The name of the file that was created should be printed, and if the `EDITOR` environment variable is set, the editor should be started on that file.

File: CHANGELOG/README.md
Comment (hanshuebner) [inline CHANGELOG/README.md:17]: List a couple of possible types to illustrate what `type` means in this context.

File: CHANGELOG/README.md
Comment (hanshuebner) [inline CHANGELOG/README.md:19]: What format is expected for this?  Describe it and enforce the format by the script (same for issues and jiras)

File: CHANGELOG/README.md
Comment (hanshuebner) [inline CHANGELOG/README.md:76]: `--type` and `--message` should not be optional.  If you don't want to prompt the user, make it an error to not provide these options.  It would then be better to change the signature to not require the options, i.e. `changelog add <type> <message>`.

File: CHANGELOG/changelog
Comment (hanshuebner) [inline CHANGELOG/changelog:21]: Add a comment that describes what CHANGELOG_PATH is going to be.

File: CHANGELOG/changelog
Comment (hanshuebner) [inline CHANGELOG/changelog:259]:
```suggestion
  print("Successfully generated release note.")
```

File: CHANGELOG/changelog
Comment (hanshuebner) [inline CHANGELOG/changelog:58]: What is the expected format of the `references` provided?  What is the `type`?

File: CHANGELOG/changelog
Comment (hanshuebner) [inline CHANGELOG/changelog:175]: Yes please.

**Review by hanshuebner — CHANGES_REQUESTED** [review]
Please add the JSON Schema validation that you've already anticipated to be needed.  More comments inline.  I like the feature and the implementation, but the usability should be enhanced.

File: CHANGELOG/changelog
Comment (vm-001) [reply CHANGELOG/changelog:83]: No, because it has been validated at the CI.

File: CHANGELOG/changelog
Comment (vm-001) [reply CHANGELOG/changelog:175]: All changelog files of the master branch have been validated by the CI `Validate changelog` job.

File: CHANGELOG/changelog
Comment (vm-001) [reply CHANGELOG/changelog:58]: The pattern is defined in the JSON schema file `CHANGELOG/schema.json`

File: CHANGELOG/README.md
Comment (vm-001) [reply CHANGELOG/README.md:76]: Setting `type` and `message` to optional does not mean they're not required, it means they're not required while creating a changelog file.

File: CHANGELOG/changelog
Comment (hanshuebner) [reply CHANGELOG/changelog:58]: Files need to be checked locally against the schema.  The documentation needs to include human readable information on the format.

File: CHANGELOG/README.md
Comment (hanshuebner) [reply CHANGELOG/README.md:76]: Please make the tool more user friendly.  If it is hard to use, developers will waste time figuring out what to do.  It should be easy to provide a tool that prevents developers from making mistakes in the first place, so please make that your goal.

File: CHANGELOG/README.md
Comment (vm-001) [reply CHANGELOG/README.md:8]: The `changelog add` command is designed to be used by GitHub workflow. Preferring keep it simple. The contributors shouldn't be distracted by adding the file. They only edit it if they are unsatisfied with the auto-generate. Before the auto CI comes up, I prefer to manually copy and edit.

Comment (hanshuebner) [reply CHANGELOG/README.md:8]: This is the first time I hear about a "GitHub" workflow.  It would help if you could provide some documentation that describes how this tool is supposed to work and be used.

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (hanshuebner) [inline .github/PULL_REQUEST_TEMPLATE.md:14]:
```suggestion
- [ ] A changelog file has been added to `CHANGELOG/unreleased/kong` (You can ask @vm-001 to do it for you)
   /or/
- [ ] The `skip-changelog` label  has been added because no changelog entry is required
```

**Review by hanshuebner — CHANGES_REQUESTED** [review]
As discussed, one change to the PR template, otherwise good to go

**Review by hanshuebner — APPROVED** [review]

File: .github/workflows/build_and_test.yml
Comment (fffonion) [inline .github/workflows/build_and_test.yml:8]:
```suggestion
    - 'CHANGELOG/**'
```

File: .github/workflows/changelog.yaml
Comment (fffonion) [inline .github/workflows/changelog.yaml:1]:
```suggestion
name: Validate Changelog
```

File: .github/workflows/changelog.yaml
Comment (fffonion) [inline .github/workflows/changelog.yaml:19]: please make sure external actions are pinned to commit sha (and version as comment), like
```
uses: tj-actions/changed-files@aaaaaaaaaaaaaaa # v37
````

File: .github/workflows/changelog.yaml
Comment (vm-001) [reply .github/workflows/changelog.yaml:1]: I will add another task in this file to auto create a changelog for PR, so I think `Changelog` is good to use

File: .github/workflows/changelog.yaml
Comment (vm-001) [reply .github/workflows/changelog.yaml]: Done

**Comment by AndyZhang0707** [comment]
Please don't merge this PR. We will merge it on Aug 29th.

**Review by windmgc — APPROVED** [review]

**Pull request merged**
